src/modules/control_allocator/ControlAllocation/ControlAllocationMultirotor.cpp

```diff
@@ -37,49 +37,30 @@
  * Control Allocation Algorithm for multirotors
  *
  * @author Julien Lecoeur <julien.lecoeur@gmail.com>
+ * @author Roman Bapst <roman@auterion.com>
  */
 
 #include "ControlAllocationMultirotor.hpp"
 
+
+
 void
 ControlAllocationMultirotor::setEffectivenessMatrix(const matrix::Matrix<float, NUM_AXES, NUM_ACTUATORS> &B)
 {
 	_B = B;
-
-	// Pseudo-inverse
-	matrix::Matrix<float, NUM_ACTUATORS, NUM_AXES> A = matrix::geninv(_B);
-
-	// Convert A to MultirotorMixer::Rotor
-	MultirotorMixer::Rotor rotors[NUM_ACTUATORS];
-
-	for (size_t i = 0; i < NUM_ACTUATORS; i++)
-	{
-		rotors[i].roll_scale = A(i, 0);
-		rotors[i].pitch_scale = A(i, 1);
-		rotors[i].yaw_scale = A(i, 2);
-		rotors[i].thrust_scale = -A(i, 5); // -Z thrust
-	}
-
-	// Update mixer
-	if (_mixer != nullptr) {
-		free(_mixer);
-	}
-
-	_mixer = new MultirotorMixer(mixer_callback, (uintptr_t)this, rotors, NUM_ACTUATORS);
+	_A = matrix::geninv(_B);
 }
 
 
 void
 ControlAllocationMultirotor::allocate()
 {
 	// Allocate
-	if (_mixer != nullptr) {
-		float outputs[NUM_ACTUATORS];
-		_mixer->mix(outputs, NUM_ACTUATORS);
-		_actuator_sp = matrix::Vector<float, NUM_ACTUATORS>(outputs);
+	_actuator_sp = _A * _control_sp;
 
-	} else {
-		_actuator_sp *= 0.0f;
+	// go through control axes from lowest to highest priority and unsaturate the actuators
+	for (unsigned i = 0; i < NUM_AXES; i++) {
+		_actuator_sp = desaturateActuators(_actuator_sp, _axis_prio_increasing[i]);
```

Comment:
Any reason why you've set the `desaturateActuators` to return _actuator_sp? Rather than getting `_actuator_sp ` as a reference, and changing the value directly inside the function?

Reply:
@junwoo091400 because of readability.